core/src/tx.rs

```diff
@@ -116,25 +116,30 @@ impl TransactionValidator {
 
         if !is_genesis {
             debug!("Validating transaction: {:?}", tx);
-
-            let AcceptedTransaction {
-                payload,
-                signatures,
-            } = tx.clone();
-            let signatures = signatures.into_iter().collect();
-
-            let signed_tx = SignedTransaction {
-                payload,
-                signatures,
-            };
-            Self::validate_with_runtime_validators(
-                &signed_tx.payload.account_id.clone(),
-                signed_tx,
-                wsv,
-            )?;
+            Self::validate_with_runtime_validator(account_id, tx.clone(), wsv)?;
         }
 
-        self.validate_and_execute_instructions(tx, wsv, is_genesis)?;
+        match tx.payload.instructions {
+            Executable::Instructions(instructions) => {
+                // Non-genesis instructions have been executed in `validate_with_runtime_validators()`.
+                //
+                // Test instructions won't be executed because of mock in
+                // `crate::validator::Validator::validate()`, so we need to do it here.
+                if is_genesis || cfg!(test) {
```

Comment:
```suggestion
                if is_genesis {
```

and copy the code in this if branch into `MockValidator::validate`. Is that possible?

Reply:
uh, now it occurred to me this is quite incorrect. Any unit tests that rely on this are in fact useless because they are not testing any actual validator/production code. Can you tell me which tests are that so I can help you figure out what to do about it?

Reply:
> and copy the code in this if branch into MockValidator::validate. Is that possible?

This is possible, but it will be ust code duplication, because it still have to be here for genesis case.

> uh, now it occurred to me this is quite incorrect. Any unit tests that rely on this are in fact useless because they are not testing any actual validator/production code. Can you tell me which tests are that so I can help you figure out what to do about it?

Well, these are all unit tests in `core` which requires transactions to be applied on wsv. They don't test validation it-self, they test other things, but I stil need this stub for tests.

> alidation should always pass for a mock validator.

Yes, `MockValidator` always returns `Ok(())`